core/store/src/trie/trie_storage.rs

```diff
@@ -162,31 +267,42 @@ impl TrieCachingStorage {
         key[8..].copy_from_slice(hash.as_ref());
         key
     }
-}
 
-impl TrieStorage for TrieCachingStorage {
-    fn retrieve_raw_bytes(&self, hash: &CryptoHash) -> Result<Vec<u8>, StorageError> {
+    pub(crate) fn retrieve_raw_bytes_with_cost(
+        &self,
+        hash: &CryptoHash,
+    ) -> Result<(Vec<u8>, RetrievalCost), StorageError> {
         let mut guard = self.cache.0.lock().expect(POISONED_LOCK_ERR);
-        if let Some(val) = guard.get(hash) {
-            Ok(val.clone())
+        if let (Some(val), cost) = guard.get_with_cost(hash) {
+            Ok((val, cost))
         } else {
             let key = Self::get_key_from_shard_uid_and_hash(self.shard_uid, hash);
             let val = self
                 .store
                 .get(ColState, key.as_ref())
                 .map_err(|_| StorageError::StorageInternalError)?;
             if let Some(val) = val {
-                if val.len() < TRIE_LIMIT_CACHED_VALUE_SIZE {
-                    guard.put(*hash, val.clone());
-                }
-                Ok(val)
+                guard.put(*hash, val.clone());
+                Ok((val, RetrievalCost::Full))
             } else {
                 // not StorageError::TrieNodeMissing because it's only for TrieMemoryPartialStorage
                 Err(StorageError::StorageInconsistentState("Trie node missing".to_string()))
             }
         }
     }
 
+    pub fn prepare_trie_cache(&self) {
+        let mut guard = self.cache.0.lock().expect(POISONED_LOCK_ERR);
+        guard.cache_state = CacheState::CachingShard;
+        guard.drain_chunk_cache();
```

Comment:
if `drain_chunk_cache` is only used here, we might want to inline the implementation to avoid confusion